File: java/client/src/main/java/com/github/fernthedev/lightchat/client/netty/ClientHandler.java
package com.github.fernthedev.lightchat.client.netty;

import com.github.fernthedev.lightchat.client.Client;
import com.github.fernthedev.lightchat.client.EventListener;
import com.github.fernthedev.lightchat.client.event.ServerConnectFinishEvent;
import com.github.fernthedev.lightchat.client.event.ServerDisconnectEvent;
import com.github.fernthedev.lightchat.core.StaticHandler;
import com.github.fernthedev.lightchat.core.packets.Packet;
import io.netty.channel.*;
import io.netty.handler.timeout.ReadTimeoutException;
import org.apache.commons.lang3.tuple.Pair;

import java.io.IOException;

@ChannelHandler.Sharable
public class ClientHandler extends ChannelInboundHandlerAdapter {

    protected EventListener listener;

    protected Client client;

    public ClientHandler(Client client, EventListener listener) {
        this.listener = listener;
        this.client = client;
    }

    /**
     * Calls {@link ChannelHandlerContext#fireChannelRegistered()} to forward
     * to the next {@link ChannelInboundHandler} in the {@link ChannelPipeline}.
     * <p>
     * Sub-classes may override this method to change behavior.
     *
     * @param ctx
     */
    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        client.getPluginManager().callEvent(new ServerConnectFinishEvent(ctx.channel()));
        super.channelActive(ctx);
    }

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        Pair<Packet, Integer> packet;



        if (msg instanceof Pair) {
            packet = (Pair<Packet, Integer>) msg;

            StaticHandler.getCore().getLogger().debug("Received the packet {} from {}", packet.getKey().getPacketName(), ctx.channel());

            listener.received(packet.getKey(), packet.getValue());

        }

        super.channelRead(ctx, msg);
    }


    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
        if (!ctx.channel().isActive()) {
            client.getLogger().info("Lost connection to server.");

            client.disconnect(ServerDisconnectEvent.DisconnectStatus.CONNECTION_LOST);
        }

        super.channelInactive(ctx);
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        if (cause instanceof ReadTimeoutException) {
            client.getLogger().info("Timed out connection");

            if (StaticHandler.isDebug()) StaticHandler.getCore().getLogger().error(cause.getMessage(), cause);

            client.disconnect(ServerDisconnectEvent.DisconnectStatus.TIMEOUT);
        } else if (cause instanceof IOException) {
            client.disconnect(ServerDisconnectEvent.DisconnectStatus.EXCEPTION);
            cause.printStackTrace();
        } else {
            cause.printStackTrace();
        }

        super.exceptionCaught(ctx, cause);
    }
}

File: java/client/src/main/java/com/github/fernthedev/lightchat/client/netty/ClientHandler.kt
package com.github.fernthedev.lightchat.client.netty

import com.github.fernthedev.lightchat.client.*
import com.github.fernthedev.lightchat.client.event.ServerConnectFinishEvent
import com.github.fernthedev.lightchat.client.event.ServerDisconnectEvent.DisconnectStatus
import com.github.fernthedev.lightchat.core.StaticHandler.core
import com.github.fernthedev.lightchat.core.StaticHandler.isDebug
import com.github.fernthedev.lightchat.core.encryption.PacketTransporter
import com.github.fernthedev.lightchat.core.packets.PacketJSON
import io.netty.channel.ChannelHandler.Sharable
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.handler.timeout.ReadTimeoutException
import kotlinx.coroutines.runBlocking
import java.io.IOException

@Sharable
open class ClientHandler(protected var client: Client, protected var listener: EventListener) :
    ChannelInboundHandlerAdapter() {
    /**
     * Calls [ChannelHandlerContext.fireChannelRegistered] to forward
     * to the next [ChannelInboundHandler] in the [ChannelPipeline].
     *
     *
     * Sub-classes may override this method to change behavior.
     *
     * @param ctx
     */
    @Throws(Exception::class)
    override fun channelActive(ctx: ChannelHandlerContext) {
        client.eventHandler.callEvent(ServerConnectFinishEvent(ctx.channel()))
        super.channelActive(ctx)
    }

    @Throws(Exception::class)
    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        if (msg is PacketTransporter) {
            core.logger.debug("Received the packet {} from {}", msg.packet.packetName, ctx.channel())
            runBlocking {
                val packet = msg.packet
                if (packet is PacketJSON) {
                    listener.received(packet, msg.id)
                }
            }
        }
        super.channelRead(ctx, msg)
    }

    /**
     * Calls [ChannelHandlerContext.fireChannelUnregistered] to forward
     * to the next [ChannelInboundHandler] in the [ChannelPipeline].
     *
     *
     * Sub-classes may override this method to change behavior.
     *
     * @param ctx
     */
    @Throws(Exception::class)
    override fun channelUnregistered(ctx: ChannelHandlerContext) {
        if (!ctx.channel().isActive) {
            client.logger.info("Lost connection to server.")
            client.disconnect(DisconnectStatus.CONNECTION_LOST)
        }
        super.channelUnregistered(ctx)
    }

    @Throws(Exception::class)
    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        when (cause) {
            is ReadTimeoutException -> {
                client.logger.info("Timed out connection")
                if (isDebug()) core.logger.error(cause.message, cause)
                client.disconnect(DisconnectStatus.TIMEOUT)
            }

            is IOException -> {
                client.disconnect(DisconnectStatus.EXCEPTION)
                cause.printStackTrace()
            }

            else -> {
                cause.printStackTrace()
            }
        }
        super.exceptionCaught(ctx, cause)
    }
}